cpp/src/neighbors/detail/knn_brute_force.cuh

```diff
@@ -519,6 +552,171 @@ void brute_force_search(
+
+  raft::detail::popc(res, filter_view, n_queries * n_dataset, nnz_view);
+  raft::copy(&nnz_h, nnz.data(), 1, stream);
+
+  raft::resource::sync_stream(res, stream);
+  float sparsity = (1.0f * nnz_h / (1.0f * n_queries * n_dataset));
+
+  if (sparsity > 0.01f) {
+    raft::resources stream_pool_handle(res);
+    raft::resource::set_cuda_stream(stream_pool_handle, stream);
+    auto idx_norm = idx.has_norms() ? const_cast<T*>(idx.norms().data_handle()) : nullptr;
+
+    tiled_brute_force_knn<T, IdxT>(stream_pool_handle,
+                                   queries.data_handle(),
+                                   idx.dataset().data_handle(),
+                                   n_queries,
+                                   n_dataset,
+                                   dim,
+                                   k,
+                                   distances.data_handle(),
+                                   neighbors.data_handle(),
+                                   metric,
+                                   2.0,
+                                   0,
+                                   0,
+                                   idx_norm,
+                                   nullptr,
+                                   filter.data());
+  } else {
+    auto csr = raft::make_device_csr_matrix<T, IdxT>(res, n_queries, n_dataset, nnz_h);
+
+    // fill csr
+    raft::sparse::convert::bitmap_to_csr(res, filter, csr);
+
+    // create filter csr view
+    auto compressed_csr_view = csr.structure_view();
+    rmm::device_uvector<IdxT> rows(compressed_csr_view.get_nnz(), stream);
+    raft::sparse::convert::csr_to_coo(compressed_csr_view.get_indptr().data(),
+                                      compressed_csr_view.get_n_rows(),
+                                      rows.data(),
+                                      compressed_csr_view.get_nnz(),
+                                      stream);
+    if (n_queries > 10) {
+      auto csr_view = raft::make_device_csr_matrix_view<T, IdxT, IdxT, IdxT>(
+        csr.get_elements().data(), compressed_csr_view);
+
+      // create dataset view
+      auto dataset_view = raft::make_device_matrix_view<const T, IdxT, raft::col_major>(
+        idx.dataset().data_handle(), dim, n_dataset);
+
+      // calc dot
+      T alpha = static_cast<T>(1.0f);
+      T beta  = static_cast<T>(0.0f);
+      raft::sparse::linalg::sddmm(res,
+                                  queries,
+                                  dataset_view,
+                                  csr_view,
+                                  raft::linalg::Operation::NON_TRANSPOSE,
+                                  raft::linalg::Operation::NON_TRANSPOSE,
+                                  raft::make_host_scalar_view<T>(&alpha),
+                                  raft::make_host_scalar_view<T>(&beta));
+    } else {
+      raft::sparse::distance::detail::faster_dot_on_csr(res,
```

Comment:
Thanks for creating an issue to fix this. Let's get this C++ side merged in. Can you please reference the github issue in a comment just abov ethis line of code so that we don't lose sight of it? Then we can get this merged in.